vspd.go

```diff
@@ -0,0 +1,543 @@
+	if err != nil {
+		return nil, err
+	}
+
+	req := SetVoteChoicesRequest{
+		Timestamp:   time.Now().Unix(),
+		TicketHash:  ticketHash,
+		VoteChoices: choices,
+	}
+
+	resp, err := v.signedVSP_HTTP("/api/v3/setvotechoices", http.MethodPost, commitmentAddr.String(), passphrase, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var setVoteChoicesResponse SetVoteChoicesResponse
+	err = json.Unmarshal(resp, &setVoteChoicesResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	err = verifyResponse(setVoteChoicesResponse.Request, req)
+	if err != nil {
+		return nil, err
+	}
+
+	data := &VspdTicketInfo{
+		Timestamp:   setVoteChoicesResponse.Timestamp,
+		Hash:        setVoteChoicesResponse.Request.TicketHash,
+		VoteChoices: setVoteChoicesResponse.Request.VoteChoices,
+	}
+
+	err = v.updateVspdDBRecord(data, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return &setVoteChoicesResponse, nil
+}
+
+// signedVSP_HTTP makes a request against a VSP API. The request will be JSON
+// encoded and signed using the provided commitment address. The signature of
+// the response is also validated using the VSP pubkey.
+func (v *VSPD) signedVSP_HTTP(url, method, commitmentAddr string, passphrase []byte, request interface{}) ([]byte, error) {
+	reqBytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	url = strings.TrimSuffix(v.baseURL, "/") + url
+	ctx := v.w.shutdownContext()
+	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	signature, err := v.w.SignMessage(passphrase, commitmentAddr, string(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("VSP-Client-Signature", base64.StdEncoding.EncodeToString(signature))
+
+	var httpClient http.Client
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("vsp responded with an error: %v", string(b))
+	}
+
+	err = validateVSPServerSignature(resp, v.pubKey, b)
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+func (v *VSPD) getTxAndAddress(ticketHash string) (*wire.MsgTx, dcrutil.Address, error) {
+	hash, err := chainhash.NewHashFromStr(ticketHash)
+	if err != nil {
+		log.Errorf("failed to retrieve hash from %s: %v", ticketHash, err)
+		return nil, nil, err
+	}
+
+	ctx := v.w.shutdownContext()
+	txs, _, err := v.w.internal.GetTransactionsByHashes(ctx, []*chainhash.Hash{hash})
+	if err != nil {
+		log.Errorf("failed to retrieve transaction for %v: %v", hash, err)
+		return nil, nil, err
+	}
+
+	commitmentAddr, err := stake.AddrFromSStxPkScrCommitment(txs[0].TxOut[1].PkScript, v.params)
+	if err != nil {
+		log.Errorf("failed to extract script addr from %v: %v", ticketHash, err)
+		return nil, nil, err
+	}
+
+	return txs[0], commitmentAddr, nil
```

Comment:
```suggestion
	if len(txs) == 0 {
		return nil, nil, fmt.Errorf("ticket has not found: %s", ticketHash)
	}

	tx := txs[0]
	ticketCommitmentOutput := tx.TxOut[1]
	commitmentAddr, err := stake.AddrFromSStxPkScrCommitment(ticketCommitmentOutput.PkScript, v.params)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to extract script addr from %v: %v", ticketHash, err)
	}

	return tx, commitmentAddr, nil
```
To make the operation here clear.